clients/src/main/java/org/apache/kafka/clients/consumer/internals/NetworkClientDelegate.java

```diff
@@ -150,7 +155,11 @@ private void maybePropagateMetadataError() {
         try {
             metadata.maybeThrowAnyException();
```

Comment:
I believe we shouldn't clear the metadata error here because once a metadata error is encountered, there might not necessarily be an event to trigger it next time. This could result in the metadata error not being propagated. Therefore, we should wait until it is thrown before clearing it.

Reply:
> This could result in the metadata error not being propagated
I see that point and I was also debating about it, but then the question is: why would we need to propagate a transient error with the subscription metadata that already disappeared?

Reply:
My observation is that when we encounter a metadata error, the application thread does not immediately execute a new event, so the background thread continues to execute `runOnce`. During the next execution of `runOnce`, the metadata error gets cleared, resulting in the event not being propagated. 
As for the event, when it throws a `TopicAuthorizationException`, it is already considered complete by `completeExceptionally`, so this event is not counted among the uncompleted ones. Make sense?

Reply:
ok, we're saying it gets cleared as soon as it's propagated (to ensure that it's indeed propagated), but we could still get an exception from a previously sent request (this was my concern). But that's what we have with the classic consumer actually, so it's consistent (there are actually tests specifically covering the behaviour considering that)
https://github.com/apache/kafka/blob/ee4264439ddda7bdebcaa845752b824abba14161/core/src/test/scala/integration/kafka/api/EndToEndAuthorizationTest.scala#L553-L554